packages/react-form-state/src/FormState.ts

```diff
@@ -185,6 +189,24 @@ export default class FormState<
     });
   }
 
+  private get clientErrors() {
+    const {fields} = this.state;
+    const clientErrors: RemoteError[] = [];
+
+    Object.keys(fields).forEach(fieldPath => {
```

Comment:
I think this will encounter problems with nested/list fields. I have been testing out function that will recursively extract the errors and field paths 👍 

```tsx
function collectErrors(
  message: string | any[] | {} | undefined,
  field?: string,
): RemoteError[] {
  if (!message) {
    return [];
  }

  if (typeof message === 'string') {
    return [{message, field: field ? [field] : undefined}];
  }

  if (Array.isArray(message)) {
    return message.flatMap(itemError => collectErrors(itemError, field));
  }

  return Object.entries(message).flatMap(([nestedField, nestedError]) =>
    collectErrors(nestedError, field ? `${field}.${nestedField}` : nestedField),
  );
}
```

and then i was building the client errors with this getter

```tsx
  private get clientErrors() {
    const {fields} = this.state;

    return Object.entries(fields)
      .flatMap(([field, {error}]) => collectErrors(error, field))
      .map(({message, field}) => ({
        field,
        message,
      }));
  }
```

Reply:
That's a good point, I was wondering about the nested/list field cases 🤔
I guess my reasoning for writing this method like I did is:
- the existing method `hasClientErrors` contains similar logic for detecting validation errors
- the type of `this.state.fields` is `FieldStates`, which I don't _think_ is shaped like a nested object or a list, but I may be mistaken

I'll try tophatting/testing this with nested/list examples and see what I find 👍

Reply:
i was testing on the consumer side with the `formData.fields` which should be similar (it looks like it's derived from `this.state.fields`). i'll give it a test in this repo to see how it works.

Reply:
🎩 in my playground is working 👍 

this is what i tested.

`FormState.tsx`
```tsx
export interface FormData<Fields> {
  fields: FieldDescriptors<Fields>;
  dirty: boolean;
  valid: boolean;
  clientErrors: RemoteError[]; // <---- added
  errors: RemoteError[];
}

  private get formData() {
    const {errors} = this.state;
    const {externalErrors = []} = this.props; // <-- added
    const {clientErrors, fields, dirty, valid} = this;

    return {
      dirty,
      valid,
      clientErrors, // <-- added
      errors: [...errors, ...externalErrors],
      fields,
    };
  }

  private get clientErrors(): RemoteError[] {
    const {fields} = this.state;

    return Object.entries(fields)
      .flatMap(([field, {error}]) => collectErrors(error, field))
      .map(({message, field}) => ({
        field,
        message,
      }));
  }

function collectErrors(
  message: string | any[] | {} | undefined,
  field?: string,
): RemoteError[] {
  if (!message) {
    return [];
  }

  if (typeof message === 'string') {
    return [{message, field: field ? [field] : undefined}];
  }

  if (Array.isArray(message)) {
    return message.flatMap(itemError => collectErrors(itemError, field));
  }

  return Object.entries(message).flatMap(([nestedField, nestedError]) =>
    collectErrors(nestedError, field ? `${field}.${nestedField}` : nestedField),
  );
}
```

`Playground` (in `web`)
```tsx
/* eslint-disable */

import * as React from 'react';
import PlaygroundProvider from 'foundation/components/PlaygroundProvider';

import {Button, TextField} from '@shopify/polaris';
import FormState, {
  validate,
  validateList,
  validateNested,
} from '@shopify/react-form-state';
import {ErrorBanner} from 'components';

export default function Playground() {
  return (
    <PlaygroundProvider>
      <FormState
        initialValues={{single: '', nested: {text: ''}, list: [{text: ''}]}}
        validators={{
          single: [
            validate(() => false, 'single error 1'),
            validate(() => false, 'single error 2'),
          ],
          nested: validateNested({text: validate(() => false, 'nested error')}),
          list: validateList({text: validate(() => false, 'list error')}),
        }}
        validateOnSubmit
      >
        {({
          clientErrors,
          dirty,
          fields: {single, nested, list},
          reset,
          submit,
        }) => (
          <div>
            {clientErrors.length > 0 && <ErrorBanner errors={clientErrors} />}
            <TextField label="Single" {...single} />
            <FormState.Nested field={nested}>
              {({text}) => <TextField label="Nested" {...text} />}
            </FormState.Nested>
            <FormState.List field={list}>
              {({text}) => <TextField label="List" {...text} />}
            </FormState.List>
            <Button onClick={reset} disabled={!dirty}>
              Reset
            </Button>
            <Button onClick={submit} disabled={!dirty}>
              Submit
            </Button>
          </div>
        )}
      </FormState>
    </PlaygroundProvider>
  );
}
```

![image](https://user-images.githubusercontent.com/5505280/54858652-5599fe80-4cc3-11e9-9bf9-c2e584dc5be7.png)

`formData.clientErrors`
![image](https://user-images.githubusercontent.com/5505280/54858717-b4f80e80-4cc3-11e9-8509-c5af794fec35.png)

**NOTE**: i needed to add `"esnext.array"` to the `tsconfig.base.json`'s `lib`

Reply:
Great! I'll add this in and write some tests 👍 Hopefully we can get this merged soon so we can deal with the dependant issues.